### DISCUSSION TITLE
Are you not supposed to use SessionFactory.currentSession?

### DISCUSSION THREAD

**Discussion by okarmazin** [quarkusio/quarkus#39231] [Q&A] [answered]
Consider the following snippet:

```kotlin
class MyController(
    private val sessionFactory: SessionFactory
) {

    @Transactional
    fun entrypoint(): Response {
        val session = sessionFactory.currentSession
        session.isDefaultReadOnly = true
        val whatever = session.use { session ->
            // ...
        }
        
        // ...
    }
```

I keep seeing `WARN: Datasource '<default>': Closing open connection prior to commit` when I do that.

Using a new session via `sessionFactory.fromSession` produces no such warning.

1. Shouldn't `sessionFactory.currentSession` return _the_ session associated with the current `@Transactional` transaction? I thought it would after reading this: https://docs.jboss.org/hibernate/orm/current/userguide/html_single/Hibernate_User_Guide.html#session-per-request
2. Is it an error to use `currentSession`? The documentation at https://quarkus.io/guides/hibernate-orm says nothing about Hibernate sessions as far as I can see.
3. Is it preferred to inject a `[Stateless]Session` instead of `SessionFactory`? The Hibernate documentation also states that too many granular sessions are considered an anti-pattern, so I guess injecting a single Session into the controller would more cleanly map to the session-per-request pattern?

[a comment by an automated account is not shown]

**Comment by yrodiere** [comment, marked as answer]
> Shouldn't sessionFactory.currentSession return the session associated with the current @Transactional transaction? I thought it would after reading this: https://docs.jboss.org/hibernate/orm/current/userguide/html_single/Hibernate_User_Guide.html#session-per-request

It could, but in Quarkus I don't think it will. It really depends on your integration. You can request that as a feature if you want: https://github.com/quarkusio/quarkus/issues/new/choose

> Is it an error to use currentSession?

It is since that feature isn't implemented.

> Is it preferred to inject a [Stateless]Session instead of SessionFactory?

Yes.

> The Hibernate documentation also states that too many granular sessions are considered an anti-pattern, so I guess injecting a single Session into the controller would more cleanly map to the session-per-request pattern?

Quarkus proxies sessions and creates one per transaction behind the scenes, which is the recommended way of doing things.

So, injecting is the way to go.

**Reply by okarmazin** [reply]
Thanks for your response, that clears things up.

> Quarkus proxies sessions and creates one per transaction behind the scenes

To make sure I understand completely, in the following piece of code, `f` and `g` work with two different sessions due to Quarkus proxying the access to the `session` field?

```
class MyController(
    private val session: Session
) {
    @GET
    fun entrypoint(): Response {
        f()
        g()
    }
    
    @Transactional
    fun f() {
        session.doSomething
    }
    
    @Transactional
    fun g() {
        session.doSomethingElse
    }
```

**Reply by yrodiere** [reply]
Assuming the session is injected (I don't see an `@Inject`?), yes they would work with different sessions.
